core/src/main/java/org/lflang/generator/docker/DockerGenerator.java

```diff
@@ -24,31 +32,174 @@ public DockerGenerator(LFGeneratorContext context) {
+      var customBuildCommands = context.getTargetConfig().get(BuildCommandsProperty.INSTANCE);
+      if (customBuildCommands != null && !customBuildCommands.isEmpty()) {
+        return customBuildCommands;
+      }
     }
-    return defaultImage();
+    return defaultBuildCommands();
+  }
+
+  /** Return the command that sources the pre-build script, if there is one. */
+  protected List<String> getPreBuildCommand() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preBuildScript();
+    if (!script.isEmpty()) {
+      return List.of("source src-gen/" + script);
+    }
+    return List.of();
+  }
+
+  /** Return the command that sources the post-build script, if there is one. */
+  protected List<String> getPostBuildCommand() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).postBuildScript();
+    if (!script.isEmpty()) {
+      return List.of("source src-gen/" + script);
+    }
+    return List.of();
+  }
+
+  /** Generate a header to print at the top of the Dockerfile. */
+  protected String generateHeader() {
+    return """
+           # Generated by the Lingua Franca compiler version %s
+           #   - Docs: https://www.lf-lang.org/docs/handbook/containerized-execution"
+           """
+        .formatted(LocalStrings.VERSION);
+  }
+
+  /** Return the Docker RUN command used for building. */
+  protected String generateRunForBuild() {
+    return "RUN "
+        + StringUtil.joinObjects(
+            Stream.of(
+                    List.of("set -ex"),
+                    getPreBuildCommand(),
+                    getBuildCommands(),
+                    getPostBuildCommand())
+                .flatMap(java.util.Collection::stream)
+                .collect(Collectors.toList()),
+            " \\\n\t&& ");
+  }
+
+  /** Return the ENTRYPOINT command. */
+  protected String generateEntryPoint() {
+    return "ENTRYPOINT ["
+        + getEntryPointCommands().stream()
+            .map(cmd -> "\"" + cmd + "\"")
+            .collect(Collectors.joining(","))
+        + "]";
+  }
+
+  /** Return a COPY command to copy the executable from the builder to the runner. */
+  protected String generateCopyOfExecutable() {
+    var lfModuleName = context.getFileConfig().name;
+    return "COPY --from=builder /lingua-franca/%s/bin/%s ./bin/%s"
+        .formatted(lfModuleName, lfModuleName, lfModuleName);
   }
 
-  public String userRunCommand() {
-    var runCmd = context.getTargetConfig().get(DockerProperty.INSTANCE).run();
-    if (runCmd != null && !runCmd.isEmpty()) {
-      return "RUN " + runCmd;
+  /** Return a COPY command to copy the scripts from the builder to the runner. */
+  protected String generateCopyOfScript() {
+    var script = context.getTargetConfig().get(DockerProperty.INSTANCE).preRunScript();
+    if (!script.isEmpty()) {
+      return "COPY --from=builder /lingua-franca/%s/src-gen/%s ./scripts/%s"
+          .formatted(context.getFileConfig().name, script, script);
```

Comment:
The script name actually could contain spaces, so we probably should account for that. The easiest way to handle that would be to crash with a runtime error and a helpful error message if the script name contains spaces.

Reply:
Addressed this in https://github.com/lf-lang/lingua-franca/pull/2234/commits/151d540279ffe8d2980936c780dec9f223c8f0f8, please check.

Reply:
I think this change is good, except that I removed the unnecessary uses of escaping for the `ENTRYPOINT` command, as we discussed on Zulip. They were causing the `DockerOptions` test to fail.